packages/devtools_app/lib/src/shared/console/primitives/scope.dart

```diff
@@ -0,0 +1,74 @@
+// Copyright 2023 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:vm_service/vm_service.dart';
+
+import '../../globals.dart';
+import '../../vm_utils.dart';
+
+class EvalScope {
+  /// Parameter `scope` for `serviceManager.service!.evaluate(...)`.
+  ///
+  /// Maps variable name to targetId.
+  Map<String, String> value({required String isolateId}) =>
+      (_refs[isolateId] ?? {}).map((key, value) => MapEntry(key, value.id!));
+
+  /// Maps isolate name to list of context variables.
+  final _refs = <String, Map<String, InstanceRef>>{};
+
+  void add(String isolateId, String variableName, InstanceRef ref) {
+    _refs.putIfAbsent(isolateId, () => {});
+    _refs[isolateId]![variableName] = ref;
+  }
+
+  /// List of variables, removed during last refresh.
+  final removedVariables = <String>[];
+
+  /// Refreshes variables in scope in response to failed eval.
+  ///
+  /// Returns true, if eval should retry.
+  /// Sets [refreshScopeChangeMessage] if scope changed.
+  Future<bool> refreshRefs(String isolateId) async {
+    removedVariables.clear();
+    final isolateItems = _refs[isolateId] ?? {};
+    var result = false;
+
+    final variableNames = [...isolateItems.keys];
+    for (final name in variableNames) {
+      final oldItem = isolateItems[name]!;
+      final refreshedItem = await _refreshRef(oldItem, isolateId);
```

Comment:
We will refresh just references to lists, not entire lists. 
I cannot test it yet, because https://github.com/dart-lang/sdk/issues/51393 is not lended yet.

In addition, yes, it is not a goal to perfectly cover every possible use case: https://docs.google.com/document/d/1M65VISBzYwiySgjKU3iM_gvcDKUzOjxGCUlQWtHOtVM/edit#bookmark=id.tmfpyez29qgr

We will do it on demand.